File: app/src/main/java/com/rakuishi/ok/view/fragment/RepoFragment.java
package com.rakuishi.ok.view.fragment;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.rakuishi.ok.R;
import com.rakuishi.ok.databinding.FragmentRepoBinding;
import com.rakuishi.ok.viewmodel.RepoViewModel;

import javax.inject.Inject;

public class RepoFragment extends BaseFragment {

    @Inject
    RepoViewModel viewModel;

    FragmentRepoBinding binding;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        getComponent().inject(this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_repo, container, false);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        binding = FragmentRepoBinding.bind(getView());
        binding.setViewModel(viewModel);
    }
}

File: app/src/main/java/com/rakuishi/ok/presentation/repo/RepoFragment.kt
package com.rakuishi.ok.presentation.repo

import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.browser.customtabs.CustomTabsIntent
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import com.rakuishi.ok.R
import com.rakuishi.ok.data.Injector
import com.rakuishi.ok.util.RequestLiveData
import kotlinx.android.synthetic.main.fragment_list.*

class RepoFragment : Fragment() {

    private val viewModel: RepoViewModel by viewModels { Injector.provideRepoViewModelFactory() }
    private val adapter: RepoAdapter = RepoAdapter()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_list, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        adapter.onItemClick =
            { CustomTabsIntent.Builder().build().launchUrl(requireContext(), Uri.parse(it.url)) }
        recyclerView.setHasFixedSize(true)
        recyclerView.adapter = adapter

        swipeRefreshLayout.setOnRefreshListener { viewModel.requestRepos() }

        subscribeUi()
    }

    private fun subscribeUi() {
        viewModel.requestLiveData.observe(viewLifecycleOwner, Observer {
            when (it.state) {
                RequestLiveData.State.LOADING -> {
                    progressBar.visibility = View.VISIBLE
                }
                RequestLiveData.State.SUCCESS -> {
                    progressBar.visibility = View.GONE
                    swipeRefreshLayout.isRefreshing = false
                    adapter.submitList(it.data)
                }
                RequestLiveData.State.FAILURE -> {
                    progressBar.visibility = View.GONE
                    swipeRefreshLayout.isRefreshing = false
                    Toast.makeText(requireContext(), it.throwable?.message, Toast.LENGTH_SHORT)
                        .show()
                }
            }
        })
    }
}
